Gives the hop a trapdoor

tralezab opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Hop has a trapdoor.

![image](https://user-images.githubusercontent.com/40974010/118064296-947de900-b34f-11eb-9074-eaf76133f00a.png)


Note: Assistants and other ne'er-do-wells may jam your trapdoor by placing lattices in the open space, jamming the closing mechanism. May want to mind that!

Groundwork is in for player built trapdoors, I'll add crafting them in another pr. I think they certainly should be held to being multi-z maps only, so i'll have to figure out disabling the crafting on non multi-z maps... and that's kind of a pain

- [ ] reapply decals

## Why It's Good For The Game

Don't like this guy? Possibly trying to climb onto your table for the nth time? Use the trapdoor. One Ha-ha-hooey later and they're outta sight.

## Changelog
:cl:
add: Hop has a trapdoor
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

tralezab committed:
ogey

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,171 @@
+	COOLDOWN_START(src, trapdoor_cooldown, trapdoor_cooldown_time)
+	internals.pulsed()
+
+/obj/item/trapdoor_remote/proc/attempt_link_up(mob/user)
```
You know what would be a lot better than this? A global signal.

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,171 @@
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
+	SIGNAL_HANDLER
+	var/turf/open/dying_trapdoor = parent
+	if(!IS_OPEN(dying_trapdoor) && !IS_OPEN(path)) //not a process of the trapdoor, so this trapdoor has been destroyed
```
probably needs to have one of the trapdoor mechanism breaking cases be if the new path is plating

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,171 @@
+
+#define TRAPDOOR_LINKING_SEARCH_RANGE 7
+
+/obj/item/trapdoor_remote
```
needs a subtype with internals already installed, this is basically what a crafted remote would be (requiring you to fill the internals before it can be used

tralezab replied:
which reminds me, this needs an attackby to insert internals and an altclick to forget the trapdoor it is linked to.

tralezab reviewed (commented):
Fumigating my code

D4C-420 commented:
images of the trap door? Where does it lead? What is preventing me from following HOP down the door once he goes in and stealing his shoes?

tralezab commented:
>images of the trap door?

It assumes the look of a normal floor.
>Where does it lead?

To the zlevel below!
>What is preventing me from following HOP down the door once he goes in and stealing his shoes?

He must construct even more trapdoors, And this is a game plan I deeply approve of

maxymax13 commented:
paramedic acess? how about ICE CHASMS ACCESS

LemonInTheDark commented:
Finally a good justification for multiz

D4C-420 commented:
> To the zlevel below!
On tram that might work but won't that send the HOP to the 9th layer of hell on icebox? seems like out of the pan into the fire.

maxymax13 commented:
you should add a room with a ladder on the z level on ice so it doesn't dump them into brazil. and on tram, what if the person doesn't have maint access? are they cucked there?

tralezab commented:
I'm running into some really annoying sdmm errors when I try to open Icebox, but I will give some sort of guarantee to people who get icebox'd that they won't be GBJ'd

tralezab commented:
### Extensive research has allowed me to discover this is where you will fall on tramstation. Seems good!
![image](https://user-images.githubusercontent.com/40974010/118056253-752a9000-b33e-11eb-8bb4-6a44780e647b.png)

D4C-420 commented:
> , what if the person doesn't have maint access?
I'm 99% sure hop has maint access along with most other heads so I doubt that's a real issue

> I will give some sort of guarantee to people who get icebox'd that they won't be GBJ'd
might I suggest a planned out path to the level 2 mining station? like a pre planned tunnel similar to the front part of lava land .

tralezab committed:
remote is more nicely placed

maxymax13 commented:
> 
> ### Extensive research has allowed me to discover this is where you will fall on tramstation. Seems good!
> 
> ![image](https://user-images.githubusercontent.com/40974010/118056253-752a9000-b33e-11eb-8bb4-6a44780e647b.png)

kinda cringe, imagine trying to do business in there and then an assistant falls on your fucking stock

tralezab commented:
that's really funny to me personally but if it's agreed to be enough of a problem i could give the maintshop marginally less space to allow for uninterrupted shoppery

D4C-420 commented:
> that's really funny to me personally but if it's agreed to be enough of a problem i could give the maintshop marginally less space to allow for uninterrupted shoppery

Maybe move it to the little cube in HOP's office ( tram has that right?) so he will instead fall just behind that wall into the red rock area?

tralezab committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into trapdoor

tralezab committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into trapdoor

LemonInTheDark commented:
> 
> 
> that's really funny to me personally but if it's agreed to be enough of a problem i could give the maintshop marginally less space to allow for uninterrupted shoppery

I think it's funny wait to see what inept has to say about it

tralezab commented:
### icebox drop location
![image](https://user-images.githubusercontent.com/40974010/118057314-91c7c780-b340-11eb-993e-ca3052fb9abd.png)

maxymax13 commented:
that's kinda cringe

tralezab commented:
ok well just get over it i guess sorry

MMMiracles commented:
replace the tram with a trap door floor

SuperNovaa41 commented:
> that's kinda cringe

Got any actual criticism or are you just going to keep saying its cringe?

maxymax13 commented:
> 
> 
> > that's kinda cringe
> 
> Got any actual criticism or are you just going to keep saying its cringe?

you don't just fucking drop people bare into the z level where you die in 20 seconds without internals or a coat that also has lavaland weather, add a room

SuperNovaa41 commented:
> > > that's kinda cringe
> > 
> > 
> > Got any actual criticism or are you just going to keep saying its cringe?
> 
> you don't just fucking drop people bare into the z level where you die in 20 seconds without internals or a coat that also has lavaland weather, add a room

Then dont go through it?

tralezab commented:
are we complaining about the trapdoor being 9 tiles away from the mining base

maxymax13 commented:
> 
> 
> > > > that's kinda cringe
> > > 
> > > 
> > > Got any actual criticism or are you just going to keep saying its cringe?
> > 
> > 
> > you don't just fucking drop people bare into the z level where you die in 20 seconds without internals or a coat that also has lavaland weather, add a room
> 
> Then dont go through it?

it's the hop that decides to send people standing in the hopline to brazil, not the people on the hopline

tralezab commented:
9 tiles gents 🍷

SuperNovaa41 commented:
> > > > > that's kinda cringe
> > > > 
> > > > 
> > > > Got any actual criticism or are you just going to keep saying its cringe?
> > > 
> > > 
> > > you don't just fucking drop people bare into the z level where you die in 20 seconds without internals or a coat that also has lavaland weather, add a room
> > 
> > 
> > Then dont go through it?
> 
> it's the hop that decides to send people standing in the hopline to brazil, not the people on the hopline

It literally takes two seconds to equip internals unless youre handicapped

maxymax13 commented:
it's almost like 1. lavaland weather will kill you and 2. you already get stunned and slowed down when you fall a z level

SuperNovaa41 commented:
> it's almost like 1. lavaland weather will kill you and 2. you already get stunned and slowed down when you fall a z level

Idk if you heard but its 9 tiles

tralezab commented:
snowstorm adjusts your body temperature by 5 to 15 degrees. lets not call it lavaland weather as that's a bit disingenuous

tralezab commented:
screenshot of the trapdoor when open for @optimumtact 
![image](https://user-images.githubusercontent.com/40974010/118063471-01907f00-b34e-11eb-8484-03955794a87b.png)

wesoda25 commented:
Dying is part of the game, who cares if 1/10 people don't manage to make it inside in time?

Mothblocks commented:
> you don't just fucking drop people bare into the z level where you die in 20 seconds without internals or a coat that also has lavaland weather, add a room

next time dont be so annoying in the hop line

Kylerace commented:
i wonder if we could extend this concept to a pool of sharks. we only need both pools and sharks for this idea

cacogen commented:
I've wanted something like this for awhile. The fun potential of multi-z has not been lived up to. You can't even jump on people anymore.

But it dumping people into the frozen tundra is pretty stupid and will make HoPs less likely to use it. Not sure what the atmosphere implications are, either. Seems like laziness not to just build a room.

I also think the HoP's trapdoor should have a sprite. So people know what they're getting into. An animation of it opening would be even better. 

I had a thought just then where player trapdoors could just be a floortile placed over open space (between plating). You step on the tile, it breaks under the weight and you fall through. That might be too easy, though. It could require being crafted with sticky tape say to hold it in place.

And if it's easy to do you need some sort of tell on the floor (similar to how wet tiles were eventually given an overlay instead of being invisible). Otherwise on multi-z, pitfall traps will both be everywhere (at least when the feature is new) and be unavoidable. I'm not sure what a logical tell would be in the situation I described. Maybe the tile could be slightly crooked, like it's not being held in place properly.

tralezab commented:
>But it dumping people into the frozen tundra is pretty stupid and will make HoPs less likely to use it. Not sure what the atmosphere implications are, either. Seems like laziness not to just build a room.

I actually checked the area and there's no weather in that area either, so the worst you can get is airlessness you can instantly use internals to fix, and a cold environment you're 9 tiles away from a building for. The mining base will look uglier with the trapdoor building, and i'd much rather have it without as long as it isn't hazardous, and due to convenient locations it just isn't that dangerous.

>I also think the HoP's trapdoor should have a sprite. So people know what they're getting into. An animation of it opening would be even better.

My idea is to make trapdoors possible to be built by anyone willing to put the time into it, and for them to assume the look of any tiled flooring in the game. So with this in mind it needs to work with no sprite... I COULD have a special panel for the hopline, but i'm kinda maybe about that

As for animations, I could actually do some kind of cool animation of the floor dropping away to the open space and I may try to make that

>I had a thought just then where player trapdoors could just be a floortile placed over open space (between plating). You step on the tile, it breaks under the weight and you fall through. That might be too easy, though. It could require being crafted with sticky tape say to hold it in place.

It's a different behavior for a tile, so i'd be a different component i'd make for this. But it isn't that bad of an idea, just crafting a makeshift trap with tape and a tile.

>And if it's easy to do you need some sort of tell on the floor (similar to how wet tiles were eventually given an overlay instead of being invisible). Otherwise on multi-z, pitfall traps will both be everywhere (at least when the feature is new) and be unavoidable. I'm not sure what a logical tell would be in the situation I described. Maybe the tile could be slightly crooked, like it's not being held in place properly.

I was thinking of adding a deep examine (when deep examines exist, normal examines tell you that you can examine further to learn something)

tralezab commented:
And actually, no wonder you've been wanting this for awhile, it was your forum post I read in minor suggestions that gave me the idea (https://tgstation13.org/phpBB/viewtopic.php?p=593744#p593744)

tralezab committed:
check commit desc, huge list

hops remotes are preloaded, mapped trapdoors are now a mapping helper, now uses global signals, remotes now have more interactions like adding and removing internals, etc

tralezab committed:
more extreme debugging

Nexofas commented:
This mechanic may open a the new meta that is HoP trapdoor griefing. 
^^
>Assign as HoP.
>Anyone in the line goes to a gulag that doesnt require the mining punishment.
>profit????
^^

Chances are this meta interpretation will be seen im the possible future rounds if this pr gets merged.Countermesures are needed against possible grief tools.

Pugie commented:
[https://www.youtube.com/watch?v=jLaCqrisEac&ab_channel=americanpresicent](url)

D4C-420 commented:
For some reason I assumed this trap door was for HOP to escape a tider not a saw trap. I'm all for death trap's but does HOP not have a mounted flash for stopping people from climbing his desk this reason?

infraredbaron commented:
Consider the following: Bluespace trap door - teleports the victim to beacon.

Mickyan commented:
Would be swell to have a pressure triggered version of this, mostly for mapping purposes

Some time ago I brought up the idea of having loose tiles in maintenance that might occasionally break when you step on them

EOBGames commented:
I've been wanting something like this for a while- anything else you want to do to it before it's merged?

TheBonded commented:
> > , what if the person doesn't have maint access?
> 
> I'm 99% sure hop has maint access along with most other heads so I doubt that's a real issue
> 
> > I will give some sort of guarantee to people who get icebox'd that they won't be GBJ'd
> 
> might I suggest a planned out path to the level 2 mining station? like a pre planned tunnel similar to the front part of lava land .

@D4C-420 the trap door isn't for the HoP to escape through, the trapdoor is in the line and allows him to push a button and drop whoever is standing on the table down. the HoP uses it to drop whoever is in the line, they don't use it to escape.

LemonInTheDark commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+		trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] tries to shut, but is jammed by [blocking]!</span>")
+		return
+	trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] swings shut!</span>")
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(trapdoor_turf_path, flags = CHANGETURF_INHERIT_AIR)
```
I'm not concerned about people falling to their frozen death, I am a bit worried about what this'll do to the hop line's atmos on icebox. Have you put any thought into that?

tralezab replied:
Not TOO much. I was thinking of sticking a heater nearby, lmao

LemonInTheDark replied:
if it's used the hop line/main hallway's gonna lock down you know

tralezab replied:
hmm. any ideas? This may be a good enough reason to stick a hut under the trapdoor.

LemonInTheDark replied:
I kinda agree with you. you could do something to like, block air, maybe something like a tile version of the atmos holofan things.

or just lean all the way into it and make the button freeze the hop line like a chad

cacogen replied:
Do you even play the game? Nobody's going to use it if it locks up that whole area with firelocks while the atmosphere is returned to normal.

LemonInTheDark reviewed (changes requested):
I love door traps

D4C-420 commented:
> @D4C-420 the trap door isn't for the HoP to escape through, the trapdoor is in the line and allows him to push a button and drop whoever is standing on the table down. the HoP uses it to drop whoever is in the line, they don't use it to escape.

ahem
> For some reason I assumed this trap door was for HOP to escape a tider not a saw trap. I'm all for death trap's but does HOP not have a mounted flash for stopping people from climbing his desk this reason?

yeah I terribly misunderstood the point of this PR. my mistake I still like the idea of HOP having a trap door for running away on Tram where his office has one door which if we compare to every other map he has at 2 doors.

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+		if(given_decals)
+			stored_decals = given_decals
+			reapply_all_decals()
+		//else
+			//record_decals()
```
I'm still working on moving decal elements over, it's not fun :(

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+
+///makes this file more legible
+#define IS_OPEN(parent) isgroundlessturf(parent)
```
See the comment, I found thinking of the trapdoor as open or not easier to read and write than checking if it was groundless.

LemonInTheDark replied:
based

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
+	SIGNAL_HANDLER
+	var/turf/open/dying_trapdoor = parent
+	if(!IS_OPEN(dying_trapdoor) && !IS_OPEN(path) && path != /turf/open/floor/plating) //not a process of the trapdoor, so this trapdoor has been destroyed
```
Changing the turf into, say, openturf is expected behavior. But turf turning from tiled flooring into plating is someone breaking the mechanism.

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+/datum/component/trapdoor/proc/try_opening()
+	var/turf/open/trapdoor_turf = parent
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(/turf/open/openspace, flags = CHANGETURF_INHERIT_AIR)
```
opening calls this

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+		try_closing()
+
+///signal called by turf changing
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
```
which calls this

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+ * change da turf my final callback. Goodbye
+ */
+/obj/item/assembly/trapdoor/proc/carry_over_trapdoor(trapdoor_turf_path, list/stored_decals, turf/new_turf)
+	new_turf.AddComponent(/datum/component/trapdoor, starts_open = FALSE, trapdoor_turf_path, src, stored_decals)
```
new turf gets the component with all the correct args, but...

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,230 @@
+	///path of the turf this should change into when the assembly is pulsed. needed for openspace trapdoors knowing what to turn back into
+	var/trapdoor_turf_path
+
+/datum/component/trapdoor/Initialize(starts_open = FALSE, trapdoor_turf_path, assembly, given_decals)
```
as far as i can tell this isn't called on the new trapdoor component. trapdoor_turf_path stays null, assembly is null as well.

tralezab reviewed (commented):
my current issue with trapdoors, if someone would like to help point out what i'm doing wrong in this process

zxaber commented:
The next step is a trap door above the supermatter.

tralezab commented:
figured it out, trapdoors work now so i'm gonna solve conflicts, add sounds, and add an animation

tralezab committed:
IT WORKS. (+sounds, matrix magic soon)

tralezab committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into trapdoor

tralezab committed:
all i can do without an in game matrix tester

tralezab committed:
tada! it doesn't work.

Mothblocks commented on code/datums/components/trapdoor.dm:
```diff
@@ -249,24 +249,38 @@
+	var/goal_scale = cos(TORADIANS(89))
+	var/goal_vertical_skew = sin(TORADIANS(89))
+	//turning clockwise requires inverting a few vars
+	if(!SIGN(horizontal_offset))
```
What do you think this is doing? Doesn't `SIGN` just give the sign? This is going to be true for anything that isnt 0, in which case SIGN wouldn't do anything.

tralezab replied:
Sign returns truthy if it is a positive, falsey otherwise

tralezab commented:
SIGN is no longer my friend

tralezab commented:
urge to save trapdoor animations for another pr: rising

tralezab committed:
committing so it exists in history

tralezab committed:
Revert "committing so it exists in history"

This reverts commit d421aa6dbc7b700b0e65c5885c23eb9aa3e043df.

tralezab committed:
map stuff back

tralezab committed:
full removal of animation, assembly icon, icemoon mapping

tralezab commented:
In other news: you can use the screwdriver to remove the internals of the remote for your own fun plans, if you so desire.

https://user-images.githubusercontent.com/40974010/118339645-4a6f4180-b4ce-11eb-8668-38285071874d.mp4

tralezab committed:
THE DECAL SOLUTION :rat:

cacogen commented:
I was thinking a tell for these could be a darker border overlay on the edge of the tile. Trapdoors usually give themselves away by the gap between them and the rest of the floor.

tralezab committed:
FINISHING THE DECAL SOLUTION :rat:

tralezab committed:
Merge branch 'master' into trapdoor

tralezab committed:
documenting

tralezab commented on code/datums/components/_component.dm:
```diff
@@ -396,6 +396,9 @@
 	var/datum/component/nt = new_type
 
+	if(istype(nt, /datum/component/trapdoor))
+		to_chat(world, "breakpoint")
```
remove this or you are an idiot

tralezab commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,265 @@
+	COOLDOWN_START(src, search_cooldown, search_cooldown_time)
+	attempt_link_up()
+
+/obj/item/trapdoor_remote/proc/attempt_link_up()
```
wrong path lol

tralezab reviewed (commented):
todo

tralezab committed:
my own review, last of the issues?

tralezab commented:
okay! ready for review/merge/hell

tralezab commented:
AFTER 5 DAYS

I DID IT

DECALS CARRY OVER

LETS GO

https://user-images.githubusercontent.com/40974010/119215206-04435080-ba81-11eb-9464-1afd26a5a409.mp4

tralezab committed:
decals carry over... my god, this sucked!

tralezab commented:
![image](https://user-images.githubusercontent.com/40974010/119215340-f17d4b80-ba81-11eb-8b15-cbbb27634dae.png)

![](https://media.discordapp.net/attachments/390728717682606081/830205866984734770/rage.gif)

tralezab committed:
lint

Mothblocks commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,269 @@
+
+/// small proc that takes passed arguments and drops it into a new element
+/datum/component/trapdoor/proc/apply_decal(description, cleanable, directional, pic)
+	parent.AddElement(/datum/element/decal, _description = description, _cleanable = cleanable, _dir = directional, _pic = pic)
```
this my friends is why we dont do underscore names and do the src trick instead

tralezab replied:
this was so fucking painful holy mother of god I AM A BELIEVER NOW

Mothblocks commented on code/datums/components/trapdoor.dm:
```diff
@@ -0,0 +1,269 @@
+///signal called by our assembly being pulsed
+/datum/component/trapdoor/proc/toggle_trapdoor(datum/source)
+	SIGNAL_HANDLER
+	if(!IS_OPEN(parent))
```
if not else lookin ass
